### PR TITLE
Bump numpy & scipy

### PR DESCRIPTION
**Pull request by nstarman**
Cherry picking from #15603.

- [x] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

### CODE REVIEW

File: tox.ini [hunk 1]
```diff
@@ -45,16 +45,18 @@ description =
...
+    numpy125: with numpy 1.25.*
     image: with image tests
     mpl334: with matplotlib 3.3.4
     mpldev: with the latest developer version of matplotlib
     double: twice in a row to check for global state changes
 
 deps =
-    numpy122: numpy==1.22.*
     numpy123: numpy==1.23.*
+    numpy123: numpy==1.24.*
```
Comment (mhvk) [inline tox.ini:58]: Copy & paste error! should be numpy124 (and next line numpy125)

**Review by mhvk — COMMENTED** [review]
Looks good modulo the two copy&paste errors. If we're updating recommended packages, should matplotlib be updated too? Or is the scipy update a consequence of the numpy one?

**Review by neutrinoceros — CHANGES_REQUESTED** [review]
LGTM, but I think we still need to update `NUMPY_TARGET_VERSION` in `setup.py` too:
https://github.com/astropy/astropy/blob/72c09473cf7e78814874c466ff2f5a11512f5b1b/setup.py#L74

File: astropy/utils/compat/numpycompat.py [hunk 2]
```diff
@@ -20,7 +19,6 @@
 # TODO: It might also be nice to have aliases to these named for specific
 # features/bugs we're checking for (ex:
 # astropy.table.table._BROKEN_UNICODE_TABLE_SORT)
-NUMPY_LT_1_22_1 = not minversion(np, "1.22.1")
 NUMPY_LT_1_23 = not minversion(np, "1.23")
```
Comment (eerovaher) [inline astropy/utils/compat/numpycompat.py:22]:
```suggestion
```

**Review by eerovaher — CHANGES_REQUESTED** [review]

**Comment by pllim** [comment]
Does this follow https://github.com/astropy/astropy-APEs/blob/main/APE18.rst ? I don't have time to check right now but we need to make sure. Thanks!

**Comment by neutrinoceros** [comment]
> Does this follow [astropy/astropy-APEs@main/APE18.rst](https://github.com/astropy/astropy-APEs/blob/main/APE18.rst?rgh-link-date=2023-12-11T16%3A29%3A45Z) ?

@pllim : close enough. `numpy<1.23` and `scipy<1.8` will very soon be legacy, following NEP 29 (followed by APE18), and we'll definitely be out of their respective support time-windows by the time 6.1 is about to be released !

**Review by neutrinoceros — APPROVED** [review]

File: astropy/units/structured.py [hunk 3]
```diff
@@ -239,8 +240,10 @@ def _recursively_apply(self, func, cls=None):
             will return a new `~astropy.units.StructuredUnit` instance.
         """
         applied = tuple(func(part) for part in self.values())
-        # Once not NUMPY_LT_1_23: results = np.void(applied, self._units.dtype).
-        results = np.array(applied, self._units.dtype)[()]
+        if sys.version_info < (3, 10):
```
Comment (mhvk) [inline astropy/units/structured.py:243]: Interesting that this is python-version dependent!  I assume you found out the hard way that python 3.9 does not work?

**Review by mhvk — APPROVED** [review]
Looks all good!

File: astropy/units/structured.py (hunk 3 above)
Comment (nstarman) [reply astropy/units/structured.py:243]: Yeah! Only the Python 3.9 test failed, complaining about void no5 accepting 2 arguments.

Comment (nstarman) [reply astropy/units/structured.py:243]: Hm. I rebased the py3.10 bump PR and found it's failing on 3.10. https://github.com/astropy/astropy/pull/15603.
So actually the `NUMPY_LT_1_23` label might be incorrect.

Comment (neutrinoceros) [reply astropy/units/structured.py:243]: According to https://numpy.org/doc/stable/reference/arrays.scalars.html#numpy.void, the second argument was added in numpy 1.24, so I guess the correct guard is `if NUMPY_LT_1_24`

Comment (nstarman) [reply astropy/units/structured.py]: Fixed!

Comment (nstarman) [reply astropy/units/structured.py]: And folded it into https://github.com/astropy/astropy/pull/15706/commits/4e49989c19fc3287b7db57edb933a1346ec3f651

Comment (mhvk) [reply astropy/units/structured.py]: OK, sorry that my comment there was so misleading!

File: astropy/units/tests/test_quantity_non_ufuncs.py [hunk 4]
```diff
@@ -2521,14 +2520,7 @@ def test_merge_arrays_str(self, flatten):
...
 untested_functions = set()
-if NUMPY_LT_1_23:
-    deprecated_functions = {
-        # Deprecated, removed in numpy 1.23
-        np.asscalar,
-        np.alen,
-    }
-else:
-    deprecated_functions = set()
+deprecated_functions = set()
```
Comment (eerovaher) [inline astropy/units/tests/test_quantity_non_ufuncs.py:2523]: Do we need `deprecated functions` here and in `astropy/utils/masked/tests/test_function_helpers.py`?

Comment (mhvk) [reply astropy/units/tests/test_quantity_non_ufuncs.py:2523]: Not strictly right now, but I suspect we will again when we really support numpy 2.0 (and 2.1?). So, I'd prefer for this to stay, but do see the case for removing it too (together with the following `untested_functions |= deprecated_functions`) -- really, the whole phrasing of this part was to help keep track of what is not tested and why.

File: astropy/units/structured.py (hunk 3 above)
Comment (nstarman) [reply astropy/units/structured.py]: It was probably planned to be done in numpy 1.23 but bumped to numpy 1.24.!

File: astropy/units/tests/test_quantity_non_ufuncs.py (hunk 4 above)
Comment (eerovaher) [reply astropy/units/tests/test_quantity_non_ufuncs.py:2523]: On one hand the lines in question would be very simple to replace if the need arises. On the other hand declaring the set of deprecated functions to be empty is very explicit. I would prefer to remove them, but I'm not going to demand it.

File: docs/changes/15706.other.rst [hunk 5]
```diff
@@ -0,0 +1 @@
+The minimum Numpy version is now 1.23 and the minimum scipy version is 1.8.
```
Comment (eerovaher) [inline docs/changes/15706.other.rst:1]: It's best to use the capitalization the projects themselves use:
```suggestion
The minimum required NumPy version is now 1.23 and the minimum required SciPy version is 1.8.
```

File: astropy/units/tests/test_quantity_non_ufuncs.py (hunk 4 above)
Comment (mhvk) [reply astropy/units/tests/test_quantity_non_ufuncs.py:2523]: Up to you, then, @nstarman!

File: astropy/utils/compat/numpycompat.py (hunk 2 above)
Comment (nstarman) [reply astropy/utils/compat/numpycompat.py]: Done.

**Review by eerovaher — APPROVED** [review]

**Comment by pllim** [comment]
@saimn or @astrofrog , is this okay to merge now or you think it is going to complicate backports?

**Comment by saimn** [comment]
Not many changes so should be fine.

**Comment by pllim** [comment]
OK, merged. Thanks, all!

**Pull request merged**